### PR TITLE
test(publish): make Jenkins lifecycle runs authoritative

### PR DESCRIPTION
**Pull request by Bojan131**
## What changed

- require exactly one selected Jenkins node and a real bearer token
- log the deployed node's status version, commit, and store backend
- validate EXPECTED_NODE_VERSION when configured, validate EXPECTED_NODE_COMMIT when configured, and require the expected commit under Jenkins
- require a confirmed positive KA id and a fresh did:dkg: UAL before counting publish success
- remove historical fallback-UAL reads, so every query proves this run's publish
- abort timed-out HTTP requests and wait for client cleanup before continuing
- stop after a timed-out publish and enforce an 11-minute in-process run deadline
- make setup, subscription, publish, and reads abortable
- checkpoint partial summaries and errors in finally
- assert the exact configured workload and calculate Grafana rates against that workload
- run every node in aggregate scripts before returning a combined failure
- make missing artifacts and database import failures exit nonzero
- make Query Remote a separate opt-in gate with V10_ENABLE_REMOTE_QUERY=true

## Meaning of a green run

By default, every configured publish, normal query, and same-node VM GET must succeed. Query Remote is not called, is not part of the completion assertion, and is written as null telemetry with remote_query_enabled=false.

If V10_ENABLE_REMOTE_QUERY=true is set, remote query is executed and becomes required again. This keeps the current release gate focused on publish, normal query, and VM GET without deleting the remote test.

## Why

The old jobs could report success while publishes timed out, could read a historical fallback UAL after a broken publish, and could lose Grafana data silently when the process or importer failed. The old remote-query/finalization race is separate from the Jenkins publish regression and must not hide successful publish/query/VM behavior.

## Required Jenkins rollout

The inline Jenkinsfiles must be updated when this test branch is adopted:

1. Set EXPECTED_NODE_COMMIT to the commit deployed on the node under test. Do not use Jenkins GIT_COMMIT, which identifies the test/publish checkout.
2. Keep V10_ENABLE_REMOTE_QUERY unset or false for this release gate.
3. Use catchError(buildResult: 'FAILURE', stageResult: 'FAILURE') so parallel nodes can finish while a failing suite keeps the job red.
4. Remove the error-swallowing suffix from summary/error database imports, while keeping imports in an always-run post or aggregation stage.

The node deployment itself must include #1650 and #1651. This PR does not modify or trigger Jenkins.

## Validation

- harness tests: 17/17 passed
- default-off subprocess proof: one confirmed publish, two local reads, zero remote requests, null remote telemetry, exit 0
- opt-in subprocess proof: remote request executed and required
- hanging publish aborts, sends no next KA, checkpoints artifacts, and exits nonzero
- invalid/missing deployment identity and publish results cannot use fallback data
- aggregate runner continues later nodes after an earlier failure, then exits nonzero
- git diff --check passed

Against the combined #1650 + #1651 six-node build:

- V10 end-to-end suite: 4/4 passed
- two consecutive Jenkins-shaped two-node runs, 10 KAs per node per run, 50 entities per KA, 12 epochs
- 40/40 confirmed publishes, 40/40 normal queries, 40/40 VM GETs
- Query Remote skipped before any call; both aggregate runs exited 0

Server-side fixes are #1650 and #1651. #1590 should not be merged because its Promise.race tail changes the authoritative publish boundary.

### CODE REVIEW

File: tests/publish/test/v10-publish-harness.test.js [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+          kaId: 17,
+          ual: 'did:dkg:base:8453/0xasset/17',
+          authorAddress: '0xpublisher',
+        }));
+        return;
+      }
+      if (req.method === 'POST' && req.url === '/api/query') {
+        localReadRequests++;
+        res.writeHead(200, { 'Content-Type': 'application/json' });
+        res.end(JSON.stringify({ result: { bindings: [{ value: 'fresh' }] } }));
```
Comment (otReviewAgent) [inline tests/publish/test/v10-publish-harness.test.js]: 🔴 Bug: Success mocks do not verify reads target the fresh publish

**What's wrong**
The added lifecycle tests claim to cover removal of fallback reads, but the fake `/api/query` and `/api/query-remote` endpoints return non-empty data for any request. That gives false confidence because the tests would pass even if production code stopped proving the current publish was readable.

**Example**
If the harness regressed to query an unscoped SPARQL string or sent `{ ual: 'did:dkg:base:8453/stale' }` to `/api/query-remote`, these tests would still return bindings/ntriples and report 100% success.

**Suggested direction**
Assert the request bodies in the fake HTTP server and return empty or an error when the fresh root/UAL is not used, so stale-data regressions fail the harness tests.

**For Agents**
In `tests/publish/test/v10-publish-harness.test.js`, make the fake server parse `/api/knowledge-assets`, `/api/query`, and `/api/query-remote` bodies. Capture the published UAL/root, then only return data when the query/VM GET targets the fresh root and remote query sends exactly the fresh UAL, expected context graph, and peer id.

File: tests/publish/src/v10-publish-lib.js [hunk 2]
```diff
@@ -362,21 +635,33 @@ export function defineChainPublishSuite(config) {
...
           console.log(`🔗 ${name} subscribing to private CG "${contextGraphId}" (syncing from curator)…`);
-          await subscribeAndWait(client, contextGraphId, name, CG_SUBSCRIBE_TIMEOUT_MS);
+          await subscribeAndWait(client, contextGraphId, name, CG_SUBSCRIBE_TIMEOUT_MS, runDeadline);
           console.log(`✅ ${name} synced CG "${contextGraphId}" — proceeding to publish`);
         }
 
         for (let i = 0; i < KA_COUNT; i++) {
+          if (i > 0 && TEST_BATCH_DELAY_MS > 0) {
+            console.log(`Waiting ${TEST_BATCH_DELAY_MS}ms before KA #${i + 1}`);
+            await sleep(TEST_BATCH_DELAY_MS);
```
Comment (otReviewAgent) [inline tests/publish/src/v10-publish-lib.js]: 🟡 Issue: Batch delay can bypass the Jenkins run deadline

**What's wrong**
The PR adds an unconditional `await sleep(TEST_BATCH_DELAY_MS)` between KAs. That wait is not capped by `RUN_TIMEOUT_MS`, even though the surrounding harness is designed to stop early so final summary/error artifacts can be written before Jenkins kills the job.

**Example**
Set `TEST_KA_BATCHES=2`, `TEST_BATCH_DELAY_MS=900000`, and `V10_RUN_TIMEOUT_MS=660000`. After KA #1 the process sleeps for 15 minutes instead of tripping the 11-minute run deadline, so Jenkins can kill the job before the harness reaches its final checkpoint/report path.

**Suggested direction**
Route the batch delay through the same run-deadline logic as HTTP operations, or explicitly fail before sleeping when the delay would exceed the remaining workload budget.

**For Agents**
In `tests/publish/src/v10-publish-lib.js`, make the inter-KA delay deadline-aware. Before sleeping, check remaining run budget; sleep at most the remaining budget and then stop/throw a `RUN_DEADLINE` error if the requested delay cannot fit. Add a harness test with a large `TEST_BATCH_DELAY_MS` proving the summary/errors files are still written before the job cap.

File: tests/publish/scripts/insert_summary_to_db.js [hunk 3]
```diff
@@ -79,12 +88,16 @@ for (const file of files) {
         console.log(`✅ Inserted ${file} into table '${tableName}'`);
     } catch (err) {
         console.error(`❌ Failed to insert ${file} into DB (table '${tableName}'):`, err.message);
+        hadFailure = true;
```
Comment (otReviewAgent) [inline tests/publish/scripts/insert_summary_to_db.js]: 🟡 Issue: Failed importer runs can leave committed partial telemetry

**What's wrong**
The PR correctly makes importer failures return a nonzero exit code, but the scripts still commit successful inserts before reporting the overall failure. That creates a partial-success state that CI/operators are likely to retry, and retries can duplicate previously committed telemetry rows.

**Example**
Run `insert_summary_to_db.js summary_SBB.json summary_DMaaST.json`: if `summary_SBB.json` inserts successfully and `summary_DMaaST.json` fails, the script exits 1 after already committing SBB. A retry of the failed Jenkins step inserts SBB again because the statement is a plain `INSERT`. The same pattern exists in `insert_errors_to_db.js` when one KA row fails and later rows/commit still proceed.

**Suggested direction**
Make the new nonzero failure behavior match the database side effect: rollback partial imports on failure, or make successful inserts idempotent before returning exit 1.

**Confidence note**
This depends on whether the telemetry tables have external de-duplication. The scripts shown here use plain INSERTs, so the retry risk exists from this code path alone.

**For Agents**
Update `tests/publish/scripts/insert_summary_to_db.js` and `tests/publish/scripts/insert_errors_to_db.js` so a failing import is atomic or idempotent. Either wrap the intended import unit in a transaction and roll it back on any `hadFailure`, or add a natural unique key/upsert for summary/error rows. Test with two files or two KA rows where the second insert fails and prove a retry cannot duplicate the first insert.

DB import failure exit codes are not regression-tested

**What's wrong**
The PR changes the importers so database failures should fail the CI step, but the added test only covers missing inputs. The most important telemetry-loss path, a valid artifact that cannot be inserted or committed, is still unverified.

**Example**
A regression that removes `hadFailure = true` from the summary insert catch would still pass `test:harness`, because the current importer test never runs with a parseable summary file and a failing DB query.

**Suggested direction**
Refactor the importers enough to inject a fake `pg.Client`, or add a controlled integration-style test, then assert DB query/commit failures set a failing exit code.

**Confidence note**
This is based on the visible harness test file: it exercises zero/missing artifact inputs but does not simulate DB connect-success followed by insert/commit failures.

**For Agents**
Add importer tests that run with valid temporary summary/error artifacts and an injectable or stubbed `pg.Client`. Cover connect success plus insert failure, and ideally commit failure, asserting the process exits nonzero for both summary and error importers.

File: tests/publish/src/v10-publish-lib.js [hunk 4]
```diff
@@ -310,19 +507,76 @@ export function defineChainPublishSuite(config) {
...
 
-        const client = makeNodeClient(hostname, token);
+        let nodeWalletAddrs = [];
+        let nodeBuild = { version: null, commit: null, storeBackend: 'unknown' };
+        let firstPayload = null;
+        let client = null;
+        let harnessError = null;
+        const runDeadline = Date.now() + RUN_TIMEOUT_MS;
+
+        try {
```
Comment (otReviewAgent) [inline tests/publish/src/v10-publish-lib.js:652]: 🟡 Issue: The PR turns the Mocha test body into the full harness engine

**What's wrong**
This concentrates several independent concepts in one deeply nested mutable closure. The behavior may be valid, but maintainability gets worse because lifecycle policy, deadline handling, telemetry, artifact checkpointing, and console reporting are now coupled through shared local variables instead of explicit inputs and outputs.

**Example**
Adding one more lifecycle operation currently means touching the counter declarations, the operation loop, failed-asset handling, completion assertion, summary object, global after-summary, and possibly error artifact formatting inside the same long closure.

**Suggested direction**
Split orchestration, per-operation execution, and artifact generation into named helpers or a small runner module. The code-judo move is to make `defineChainPublishSuite` register tests, not own every harness concern.

**For Agents**
In `tests/publish/src/v10-publish-lib.js`, preserve the current publish/query/VM/optional-remote behavior but extract a node-runner boundary like `runNodeLifecycle(...) -> RunResult`, plus separate helpers for `validateHarnessConfig`, `prepareContextGraph`, and `writeRunArtifacts`. The Mocha `it` should mostly select a node, call the runner, checkpoint the result, and assert completion. Existing harness tests should still prove timeout, missing UAL, wrong build, and remote opt-in behavior.

File: tests/publish/test/v10-publish-harness.test.js [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+        'Alpha',
+        1000,
+        Date.now() + 15,
+      ),
+      (error) => error.code === 'RUN_DEADLINE' && error.runDeadline === true,
+    );
+    assert.equal(abortObserved, true);
+  });
+
+  it('defaults remote verification off, makes no remote request, and reports null telemetry', async () => {
```
Comment (otReviewAgent) [inline tests/publish/test/v10-publish-harness.test.js]: 🟡 Issue: Extract the repeated integration-test harness fixture

**What's wrong**
The new test file is mostly duplicated harness scaffolding. That makes the scenarios harder to scan and increases the chance that future harness changes update one test path but leave the others using stale env defaults or cleanup behavior.

**Example**
If the child-process timeout, common env defaults, or summary filename changes, the same plumbing has to be updated across five separate integration tests, while the actual scenario-specific differences are only the route responses and a few env overrides.

**Suggested direction**
Move the server setup, child Mocha execution, output capture, timeout handling, summary/error artifact reads, and cleanup into reusable helpers. Keep each test focused on its scenario.

**For Agents**
In `tests/publish/test/v10-publish-harness.test.js`, preserve the same scenario coverage but introduce helpers such as `withMockNodeServer(routes, fn)` and `runPublishSpec({ env, cwd, timeoutMs })`. Consider table-driving the success/default-remote/timeout/missing-UAL/wrong-build scenarios so each test only states the mocked responses and expected assertions.

File: tests/publish/scripts/run_node_suites.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import { spawnSync } from 'child_process';
+import { pathToFileURL } from 'url';
+
+export const NODE_SUITE_GROUPS = {
```
Comment (otReviewAgent) [inline tests/publish/scripts/run_node_suites.js]: 🟡 Issue: Avoid duplicating suite topology between package scripts and the runner

**What's wrong**
The aggregate runner introduces another source of truth for which node suites belong to each chain. The implementation is small, but the design is brittle because the important topology now lives as duplicated strings across files rather than one explicit model.

**Example**
Adding a new Base mainnet node would require updating the spec's node list, adding a `test:base:<node>:mainnet` package script, and also adding that exact string to `NODE_SUITE_GROUPS`. Missing the last step leaves `npm run test:base:mainnet` silently incomplete.

**Suggested direction**
Make `run_node_suites.js` the canonical suite orchestrator instead of calling back into duplicated npm script names, or move the shared topology into a manifest used by both aggregate and single-node commands.

**Confidence note**
This is a maintainability concern from the diff structure; it does not imply the current group contents are wrong.

**For Agents**
Replace the stringly `NODE_SUITE_GROUPS` map with one canonical suite manifest that includes spec path, node names, and report filenames, then have the runner spawn Mocha directly with `NODE_TO_TEST` and reporter options. Keep package.json as thin entry points into that manifest.

File: tests/publish/src/v10-publish-lib.js (hunk 2 above)
Comment (zsculac) [reply tests/publish/src/v10-publish-lib.js]: Independent verification and acceptance criteria for the implementing agent:

This is a **conditional deadline hole**, not evidence that the current Jenkins job is already sleeping. `TEST_BATCH_DELAY_MS` defaults to `0`, and no checked-in configuration sets it. The problem appears only when Jenkins or another caller supplies a non-zero value. This PR newly makes the hardened chain suite honor that knob while also promising an absolute in-process deadline.

The raw `await sleep(TEST_BATCH_DELAY_MS)` is outside the deadline helper. For example, with two batches, `V10_RUN_TIMEOUT_MS=2000`, and `TEST_BATCH_DELAY_MS=5000`, KA 1 can finish and the process can then sleep past the run deadline. An outer Jenkins timeout may kill it before the `finally` block writes `summary_*.json` and `errors_*.json`.

Please make the delay deadline-aware. Before sleeping, calculate `runDeadline - Date.now()`. If the requested delay does not fit, either fail immediately with the same `RUN_DEADLINE` shape or wait only until the deadline and then throw. In both cases, control must reach the existing `finally` block. Add a subprocess regression test using two batches, a short run timeout, and a longer batch delay; assert non-zero exit plus both checkpoint files. Preserve the current zero-delay behavior.

File: tests/publish/test/v10-publish-harness.test.js (hunk 1 above)
Comment (zsculac) [reply tests/publish/test/v10-publish-harness.test.js]: Independent verification and concrete acceptance criteria for the implementing agent:

The runtime code is currently doing the right thing: local reads build SPARQL from the newly generated `rootEntity`, and remote reads send the UAL returned by that publish. The gap is specifically in the regression proof. The success fixtures increment counters and return non-empty data for every `/api/query` and `/api/query-remote` body, so they would stay green if a later edit sent an unscoped query, the wrong context graph, or a stale UAL.

Please make the mock server consume and parse request JSON. From the publish request, capture the unique generated root from `quads`; use a fixed fresh UAL in the publish response. Return local bindings only when both `/api/query` calls contain that exact root, the expected `contextGraphId`, and `view: verifiable-memory`. Return remote triples only when `peerId`, `contextGraphId`, `lookupType: ENTITY_BY_UAL`, and `ual` exactly match the expected fresh values.

Also add at least one negative regression case: an unscoped local query or stale remote UAL must produce an empty/error response and make the child lifecycle process exit non-zero. This comment is about preventing future false confidence, not claiming that the present request construction is broken.

**Comment by zsculac** [comment]
### CI coverage for the new harness

I ran the exact PR head (`48f263442`) under the repository Node version and `npm run test:harness` passed 17/17. The remaining problem is that GitHub does not independently run that command:

- this PR targets `test/publish`, while `.github/workflows/ci.yml` only triggers for pull requests targeting `main`;
- `tests/publish` is not included in `pnpm-workspace.yaml`, so the root Turbo test job would not discover this package even if the workflow triggered;
- inside `tests/publish/package.json`, plain `npm test` still runs the older generic lifecycle spec, not `test:harness`.

That leaves the main proof for a test-harness PR as a manual claim. Please add a dedicated CI job or workflow for PRs targeting `test/publish`. It should use Node 22 from `.nvmrc`, set `working-directory: tests/publish`, run `npm ci`, then run `npm run test:harness`. This suite uses local mock HTTP servers and does not need live node credentials or chain access.

Acceptance criteria for the implementing agent:

1. pushing a change to the harness automatically produces a visible Harness Test check on this PR;
2. intentionally breaking one harness assertion makes that check fail;
3. the job runs the exact locked dependencies from `tests/publish/package-lock.json`;
4. either make `npm test` include the harness or keep the CI command explicit and document that the default test script is the legacy suite.

If branch protection is available for `test/publish`, make this check required so the alarm system is not guarded by an honor-system sticky note.

File: tests/publish/scripts/insert_summary_to_db.js (hunk 3 above)
Comment (otReviewAgent) [reply tests/publish/scripts/insert_summary_to_db.js]: 🔴 Bug: Importers now reject valid mixed-network artifact batches

**What's wrong**
The changed importer treats a batch containing both mainnet and testnet artifacts as an error. The old script processed each file independently and selected the correct DB host/table per file. If Jenkins or an operator imports all generated artifacts with one glob, the whole import now fails and no telemetry is persisted.

**Example**
`node scripts/insert_summary_to_db.js summary_Base_Mainnet.json summary_Base_Testnet.json` now exits 1 before connecting, even though both files are valid and could be inserted into their respective tables. The same applies to `insert_errors_to_db.js` with `errors_*` files from both networks.

**Suggested direction**
Group artifacts by target and commit each target group atomically, or ensure every existing caller is changed to invoke the importer separately for mainnet and testnet artifacts.

**Confidence note**
The script-level behavior changed from per-file routing to batch rejection; I did not find a caller in this repo that proves mixed-network artifacts are never passed together.

**For Agents**
Look at `parseSummaries` and `parseErrorArtifacts`. Preserve the new all-or-nothing behavior within a single DB/table target, but group artifacts by `isMainnet` or full target and run one transaction per group instead of rejecting mixed valid inputs. Add a regression that passes one mainnet and one testnet artifact in a single invocation and proves both groups import, or explicitly update callers to split invocations before this check.

Mixed mainnet/testnet import rejection is untested

**What's wrong**
The PR adds a guard that rejects mixed mainnet/testnet artifact batches before opening one transaction, but the new tests only exercise same-target batches and database failures. That leaves a data-integrity branch unverified: a future change could silently route mixed artifacts into one table/host and still keep the CI harness green.

**Example**
A regression that removed the `targetKinds.size > 1` check would still pass the current harness tests, even though `importSummaries([mainnetSummary, testnetSummary])` should return `1`, avoid connecting to Postgres, and log the mixed-batch error.

**Suggested direction**
Add explicit regression tests for mixed-target summary and error batches so the new atomicity guard is verified, not just the transaction rollback path.

**For Agents**
Add harness coverage near `rolls back summary and error batches on insert or commit failure` that builds mixed mainnet/testnet summary and error inputs, asserts `importSummaries`/`importErrors` return `1`, and asserts the fake client factory is not called. Preserve the existing rollback tests for single-target batches.

File: tests/publish/src/v10-publish-lib.js (hunk 4 above)
Comment (otReviewAgent) [reply tests/publish/src/v10-publish-lib.js:652]: 🟡 Issue: Split the harness utilities before `v10-publish-lib.js` becomes a catch-all module

**What's wrong**
This change concentrates several unrelated concepts in one already-large file: timeout primitives, Jenkins policy, node validation, artifact serialization, HTTP client shape, context graph setup, lifecycle orchestration, and final console reporting. That makes the file a dumping ground for future harness behavior and pushes it right up against the 1k-line decomposition threshold. The exports also expose implementation details mainly so the new tests can reach them, which hardens the monolith instead of creating clear ownership boundaries.

**Example**
A reader changing publish retry behavior now has to scan past Jenkins build validation, token validation, DB artifact shapes, summary file creation, and Mocha suite registration in the same file. The harness tests also import many internals directly from this catch-all file, which makes it harder to split later without touching unrelated test cases.

**Suggested direction**
Move the newly exported utility clusters into focused modules and leave `v10-publish-lib.js` to compose the suite. This would delete a lot of incidental coupling without changing behavior.

**For Agents**
In `tests/publish/src/v10-publish-lib.js`, preserve the lifecycle behavior but split the new responsibilities into focused modules, for example `run-deadline.js`, `preflight-validation.js`, `artifact-checkpoint.js`, and `node-client.js`. Keep `defineChainPublishSuite` as the orchestration layer and update `tests/publish/test/v10-publish-harness.test.js` to import helpers from their owning modules. Existing harness tests should continue to prove timeout cancellation, validation, artifact checkpointing, and lifecycle completion.

File: tests/publish/scripts/insert_summary_to_db.js (hunk 3 above)
Comment (otReviewAgent) [reply tests/publish/scripts/insert_summary_to_db.js]: 🟡 Issue: Collapse the duplicated DB batch transaction flow into one importer abstraction

**What's wrong**
The PR improves each importer locally, but it does so by copying the same transaction lifecycle into both scripts. That is a structural missed opportunity: the atomic batch behavior is the new core abstraction, while summary-vs-error is just different parsing and row mapping. Duplicating the orchestration leaves two places for rollback, close, timeout, and logging policy to diverge.

**Example**
The rollback/close policy now exists twice. If the team later changes timeout settings, retries, close-error handling, or atomicity rules for Jenkins imports, both scripts need the same edit or their failure behavior will drift.

**Suggested direction**
Introduce a small shared transaction runner for Jenkins artifact imports, with summary/error scripts supplying only parsing and insert mapping. That keeps the atomicity policy in one canonical place.

**Confidence note**
This is based on the duplicated structure visible in the diff; there may be an existing shared scripts helper outside the changed files, but the new code does not use one.

**For Agents**
Look at `tests/publish/scripts/insert_summary_to_db.js` and `tests/publish/scripts/insert_errors_to_db.js`. Preserve current CLI behavior and row shapes, but extract a shared batch transaction helper such as `importArtifactBatch({ files, parseArtifacts, createClient, insertRows, logger })`. Keep importer-specific tests for summary/error parsing plus shared tests for connect failure, insert failure, commit failure, rollback, and close handling.

File: tests/publish/src/v10-publish-lib.js (hunk 4 above)
Comment (otReviewAgent) [reply tests/publish/src/v10-publish-lib.js:652]: 🟡 Issue: This pushes the publish harness past 1k lines instead of decomposing it

**What's wrong**
The PR adds enough responsibility to push an already-large harness file over 1000 lines. The result is harder to scan and harder to change safely because unrelated concepts now live in one module. This is exactly the kind of growth that should trigger decomposition first.

**Example**
A change to summary artifact shape now has to happen in the same 1k+ line module that also owns HTTP cancellation, publish/query orchestration, remote peer lookup, and Mocha suite registration. That is a sign the PR should split responsibilities before adding more harness behavior here.

**Suggested direction**
Move the newly added pure helpers and artifact-writing logic out of `v10-publish-lib.js` into focused modules, leaving this file as the suite registration/orchestration layer. This should delete concepts from the main file rather than just rearranging them.

**For Agents**
In `tests/publish/src/v10-publish-lib.js`, preserve the current harness behavior but extract focused modules before merging: e.g. run/deadline helpers, node validation, node client, artifact checkpointing, and lifecycle orchestration. Keep the spec-facing API small and run `npm run test:harness` afterward.

File: tests/publish/scripts/insert_summary_to_db.js (hunk 3 above)
Comment (otReviewAgent) [reply tests/publish/scripts/insert_summary_to_db.js]: 🟡 Issue: The DB importers duplicate the same transaction state machine

**What's wrong**
The PR improves atomic imports, but implements the atomic transaction orchestration twice. That duplication makes the reliability policy harder to maintain because future fixes to commit, rollback, close, or return-code behavior must be mirrored manually across both scripts.

**Example**
If rollback or socket-close handling changes, the same transaction policy has to be edited in both `insert_summary_to_db.js` and `insert_errors_to_db.js`. The summary importer already carries a detailed close-failure rationale while the error importer has a shorter variant, which shows how these duplicated paths can drift.

**Suggested direction**
Create one shared transaction/batch-import helper that accepts the target connection info, insert query, and row/value iterator. Let each importer focus only on parsing its artifact shape and mapping rows.

**For Agents**
Extract a shared DB import helper under `tests/publish/scripts/`, such as `withDbTransaction` or `importArtifactBatch`. Keep summary/error parsing and row mapping local, but centralize client config, BEGIN/COMMIT/ROLLBACK/end handling, return-code policy, and logging hooks. Preserve current atomic behavior and importer tests.

Comment (otReviewAgent) [reply tests/publish/scripts/insert_summary_to_db.js]: 🟡 Issue: Mixed-target import rejection lacks regression coverage

**What's wrong**
The importers now depend on a preflight guard to prevent a mixed mainnet/testnet batch from being inserted into the first artifact’s table. Current tests cover rollback behavior but not this guard, so a regression could still pass while allowing telemetry rows into the wrong table.

**Example**
`importSummaries([mainnetSummary, testnetSummary], fakeClientFactory)` should return `1` and never connect. The same should be proven for `importErrors` with one mainnet artifact and one testnet artifact.

**Suggested direction**
Cover the new mixed-batch validation for both summary and error importers, including the important behavior that rejection happens before opening a DB transaction.

**For Agents**
Add a harness test that builds mixed mainnet/testnet summary files and mixed error artifacts, calls both importers with a fake client factory that records whether it was created/connected, and asserts the import returns `1` before any DB work.

File: tests/publish/scripts/insert_summary_to_db.js [hunk 7]
```diff
@@ -62,29 +91,56 @@ for (const file of files) {
...
-            summary.non_publisher_get_success_rate,
-            summary.average_publish_time,
-            summary.average_query_time,
-            summary.average_publisher_get_time,
-            summary.average_non_publisher_get_time,
-            summary.time_stamp,
-        ]);
+        for (const { file, summary } of artifacts) {
+            try {
+                await db.query(query, [
```
Comment (otReviewAgent) [inline tests/publish/scripts/insert_summary_to_db.js:96]: 🔴 Bug: DB import drops the new SWM cleanup failure signal

**What's wrong**
This change adds a new blocking lifecycle check, but the importers still persist only the old telemetry shape. A run can fail solely because SWM cleanup failed while the imported summary records publish/query/VM GET as 100% and drops `harness_error`; the detailed SWM error is also not routed into any DB error column. That loses the failure state at the reporting boundary.

**Example**
If publish, query, and VM GET succeed but the new SWM cleanup check fails, the JSON summary contains `publish_success_rate: "100.00"`, `query_success_rate: "100.00"`, `publisher_get_success_rate: "100.00"`, `shared_working_memory_cleanup_success_rate: "0.00"`, and a non-null `harness_error`. The importer persists only the three 100% old columns, so downstream DB/Grafana consumers see a green run with no persisted SWM failure signal.

**Suggested direction**
Make the DB import contract match the new lifecycle gate, or reject summaries/errors whose failure state cannot be represented safely.

**For Agents**
Update the summary and error import path together. In `insert_summary_to_db.js`, persist `harness_error` and the SWM cleanup result if the DB supports new columns, or fail the import when a summary has unrepresentable failure state. In `insert_errors_to_db.js`, route `SWM cleanup` detailed keys to an explicit column/status or fail instead of inserting blank error rows. Add an importer test with a SWM-cleanup-only failure proving the persisted telemetry cannot look green.

File: tests/publish/src/v10-helpers.js [hunk 8]
```diff
@@ -202,10 +202,52 @@ export async function httpCreateContextGraph(id, name, description) {
...
+ * HTTP 207 and any response error field must therefore never count as a clean
+ * publish in Jenkins telemetry.
+ */
+export function requireCompleteLifecyclePublishResponse(httpStatus, data = {}) {
+  const responseErrors = [...new Set([
+    ...lifecycleErrorMessages(data.errors),
+    ...lifecycleErrorMessages(data.error),
+    ...lifecycleErrorMessages(data.contextGraphError),
+  ])];
+  if (httpStatus !== 207 && responseErrors.length === 0) return data;
```
Comment (otReviewAgent) [inline tests/publish/src/v10-helpers.js:233]: 🟡 Issue: Bare HTTP 207 rejection is not covered

**What's wrong**
The new behavior says any HTTP 207 must never count as a clean publish, but the tests only prove rejection when 207 is accompanied by error fields. That leaves the status-only contract unverified.

**Example**
A regression that changed the guard to `if (responseErrors.length === 0) return data` would still pass the current tests, because the only 207 test also supplies `errors`/`contextGraphError`. The missing case is `requireCompleteLifecyclePublishResponse(207, { status: 'vm-confirmed', errors: [] })`, which should throw.

**Suggested direction**
Add a regression test for a 207 response with an otherwise clean body, because the changed contract is that the status code alone means partial lifecycle completion.

**For Agents**
In `tests/publish/test/v10-publish-harness.test.js`, add a unit assertion and/or mock-node run where publish returns HTTP 207 with no `errors`, `error`, or `contextGraphError`; prove the helper throws and the harness exits nonzero without counting publish success.
